src/Data/Text/Internal/IO.hs

```diff
@@ -168,17 +175,41 @@ readChunk hh@Handle__{..} buf = do
 
 -- | Print a @Stream Char@.
 hPutStream :: Handle -> Stream Char -> IO ()
--- This function is lifted almost verbatim from GHC.IO.Handle.Text.
-hPutStream h str = do
-  (buffer_mode, nl) <-
+hPutStream h str = hPutStreamOrUtf8 h str Nothing
+
+-- | Write a string to a handle.
+hPutStr :: Handle -> Text -> IO ()
+hPutStr h t = hPutStreamOrUtf8 h (stream t) (Just putUtf8)
+  where putUtf8 = B.hPutStr h (encodeUtf8 t)
+
+-- | Write a string to a handle, followed by a newline.
+hPutStrLn :: Handle -> Text -> IO ()
+hPutStrLn h t = hPutStreamOrUtf8 h (streamLn t) (Just putUtf8)
+  where putUtf8 = hPutBuilder h (encodeUtf8Builder t <> charUtf8 '\n')
```

Comment:
Why not `B.hPutStrLn h (encodeUtf8 t)`?

Reply:
It would have to be `B.hPutStrLn h (encodeUtf8 (t <> '\n'))`. I couldn't find another way to append the `'\n'` at the end in constant time. Open to suggestions.

Reply:
I'd expect `B.hPutStrLn` to append `'\n'` on its own, am I missing something?

Reply:
Oh I didn't see your `Ln`. There is no `B.hPutStrLn`... right?

Reply:
```haskell
hPutStrLn :: Handle -> ByteString -> IO ()
hPutStrLn h ps = hPut h ps >> hPut h (L.singleton 0x0a)
```

https://github.com/haskell/bytestring/blob/46a3aeb179c26d2385f95ba518eaa5464c94ceaa/Data/ByteString/Lazy/Char8.hs#L932

Reply:
I'd say that non-100%-atomicity of `B.hPutStrLn` is for `bytestring` to resolve, not for us. But does `hPutBuilder` guarantee atomic printing?

Reply:
`B.hPutBuilder` does if it can fit in the buffer.

Reply:
I reopened https://github.com/haskell/bytestring/issues/200 and added a comment.